File: cardroom/briscola/game/env.py
import pygame

from copy import deepcopy
from cardroom.briscola.game.dealer import BriscolaDealer
from cardroom.briscola.game.player import BriscolaPlayer
from cardroom.briscola.agents.random import RandomAgent
from cardroom.briscola.agents.human import HumanAgent
from cardroom.briscola.render.pygame import BriscolaPygame

RENDER_MODES = [None, "text", "pygame"]

class BriscolaEnv:
    """Environment for a game of Briscola."""
    def __init__(self, names: list[str,str]|None, render_mode: str = "text"):
        """Instantiates the environment.
        Args:
            names: names of players who are going to play the game.
            render_mode: render mode for the game.
        """
        self.n_players = 2
        if render_mode not in RENDER_MODES:
            raise NotImplementedError(f"Render mode {render_mode} not implented.")
        self.render_mode = render_mode
        self.instantiate_players(names)
        self.reset()
        self.render()

    def instantiate_players(self, names: list[str, str] | None = None):
        """Instantiates the players for the game.
        Args:
            players: list of players who are going to play the game.
        """
        if names is None:
            self.players = [BriscolaPlayer(f"P{i}") for i in range(self.n_players)]
        else:
            if len(names) != self.n_players:
                raise ValueError(f"Expected {self.n_players} players, got {len(names)}")
            self.players = [BriscolaPlayer(name) for name in names]

    def set_turn_order(self, first_id: int):
        """Sets the turn order for the players given who has to start.
        Args:
            first_id: index of the player who has to start.
        """
        if first_id < 0 or first_id >= self.n_players:
            raise ValueError(f"Invalid player index {first_id}, must be in [0, {self.n_players - 1}]")
        default_order = list(range(self.n_players))
        self.turn_order = default_order[first_id:] + default_order[:first_id]


    def reset(self):
        """Resets the env to the initial state at the beginning of a new game."""
        self.done = False
        self.played_cards_history = []
        self.turn_history = []
        
        # clean up players and dealer
        self.dealer = BriscolaDealer()
        for player in self.players:
            player.reset()
        
        # set up new game
        self.briscola_spy = self.dealer.spy
        self.briscola_suit_id = self.briscola_spy.suit_id
        self.current_player_id = self.dealer.draw_starting_player(self.n_players)
        self.set_turn_order(self.current_player_id)
        # deal first 3 cards
        for id in self.turn_order:
            player = self.players[id]
            for _ in range(3):
                player.hand.append(self.dealer.deal())
        self.turn_counter = 1
        self.cards_on_table = []

    def pass_turn(self):
        """Passes the turn to the next player."""
        self.current_player_id = (self.current_player_id + 1) % self.n_players

    def resolve_turn(self):
        """Updates the game after a whole turn has been played."""
        if len(self.cards_on_table) != self.n_players:
            raise ValueError(f"Cannot resolve turn, expected {self.n_players} cards on table but got {len(self.cards_on_table)}.")
        # first card suit commands, without a briscola
        first_suit_id = self.cards_on_table[0].suit_id
        # judge winner and update points
        played_rank = []
        played_points = 0
        for card in self.cards_on_table:
            effective_rank = card.rank
            if card.suit_id == first_suit_id:
                effective_rank += 50
            played_rank.append(effective_rank)
            played_points += card.points
        winning_card_idx = played_rank.index(max(played_rank))
        winning_player_id = self.turn_order[winning_card_idx]
        self.players[winning_player_id].points += played_points
        self.players[winning_player_id].taken_cards.extend(self.cards_on_table)
        # update turn order
        self.current_player_id = winning_player_id
        self.set_turn_order(winning_player_id)
        # reset cards on table
        self.cards_on_table = []

    def get_state(self) -> dict:
        """Retrieves the current game state.
        Returns:
            state: current game state.
        """
        state = {}
        state["player_names"] = [player.name for player in self.players]
        state["current_player"] = self.current_player_id
        state["turn_order"] = self.turn_order
        state["turn_counter"] = self.turn_counter
        state["deck_left"] = self.dealer.cards_left
        state["cards_on_table"] = self.cards_on_table
        state["briscola_id"] = self.briscola_suit_id
        state["briscola_spy"] = self.briscola_spy
        state["hands"] = [player.hand for player in self.players]
        state["points"] = [player.points for player in self.players]
        state["taken_cards"] = [player.taken_cards for player in self.players]
        state["all_played_cards"] = self.played_cards_history
        state["turn_history"] = self.turn_history
        state["done"] = self.done

        return state
    
    def get_observation(self) -> dict:
        """Retrieves observation of game state for current player.
        Returns:
            obs: observation of the game state for the current player.
        """
        state = self.get_state().copy()
        obs = {}

        # hide other player hand from player obs
        for key in state:
            if key in ["hands", "points"]:
                continue
            else:
                obs[key] = state[key]

        obs["hand"] = state["hands"][self.current_player_id]
        obs["current_player_points"] = state["points"].pop(self.current_player_id)
        obs["other_player_points"] = state["points"].pop()
        return obs

    def step(self, action: int):
        """Executes the action of the current player and updates the state of the game.
        Args:
            action: index of the card to be played by the current player."""
        # check if current player if last to play
        if self.current_player_id == self.turn_order[-1]:
            is_last = True
        else:
            is_last = False
        # get current player
        player = self.players[self.current_player_id]
        # play his action
        card = player.play_card(action)
        self.cards_on_table.append(card)
        self.played_cards_history.append(card)
        self.turn_history.append(self.current_player_id)

        # if he's last, resolve the turn, else just pass the turn
        if not is_last:
            self.pass_turn()
        else:
            self.resolve_turn()
            self.turn_counter += 1
            # check if game is over, else deal new cards
            if self.turn_counter > 20:
                self.result, points = self.check_result()
                self.done = True
            else:
                if self.dealer.cards_left > 0:
                    for id in self.turn_order:
                        player = self.players[id]
                        player.hand.append(self.dealer.deal())
        self.render()

    def check_result(self) -> tuple[int, list[int, int]]:
        """Checks result of the game.
        Returns:
            result : 0 if a draw, 1 if p0 won, -1 if p1 won.
            points: list of points of the players.
        """
        points = []
        for player in self.players:
            points.append(player.points)
        if points[0] == points[1]:
            result = 0
        elif points[0] > points[1]:
            result = 1
        else:
            result = -1
        result = result
        return result, points
    
    def get_legal_actions(self) -> list:
        """Returns legal actions at the current state of the game.
        Returns:
            legal_actions: list of legal actions for the current player.
        """
        obs = self.get_observation()
        return list(range(len(obs["hand"])))

    def render(self):
        """Renders the current state of the game."""
        state = self.get_state()

        if self.render_mode is None:
            return # don't render
        elif self.render_mode == "text":
            self.text_render(state)
        elif self.render_mode == "pygame":
            self.pygame_render(state)

    def text_render(self, state):
        """Renders in text mode.
        Args:
            state: current state of the game.
        """
        # print beginning of the game
        if state["turn_counter"] == 1 and len(state["cards_on_table"]) == 0:
            print("New game starting!")
            print(f"Players: {[player.name for player in self.players]}")
            print(f"{self.players[self.current_player_id].name} to start.")
            print(f"Briscola: {self.briscola_spy.suit} - Spy: {self.briscola_spy}")
            self.last_rendered_turn = 1
            return
        # print turn start
        if len(state["cards_on_table"]) == 1:
            print(state["points"])
            print(f"---TURN {state['turn_counter']}---")
        # print last action
        print(f"{self.players[state['turn_history'][-1]].name} plays {state['all_played_cards'][-1]}.")
        # print end of the game
        if state["done"]:
            print("Game finished.")
            for player in self.players:
                print(f"{player.name}: {player.points} points.")
            if state["points"][0] == state["points"][1]:
                print("It's a draw!")
            elif state["points"][0] > state["points"][1]:
                print(f"{self.players[0].name} wins!")
            else:
                print(f"{self.players[1].name} wins!")

    def pygame_render(self, state):
        """Renders in pygame mode.
        Args:
            state: current state of the game.
        """
        # also render text for logging
        self.text_render(state)
        
        # init pygame
        if not pygame.get_init():
            self.pygame = BriscolaPygame()
            # for player in self.players:
            #     #if isinstance(player.agent, HumanAgent) and player.agent.input_mode ==  "pygame":
            #     # TODO: at the moment only humans players are handled
            #     player.agent.set_pygame_action_retriever(self.pygame)

        if self.pygame.running:
            # self.pygame.handle_events() is called by human agent select_action
            self.pygame.draw(state)
        else:
            self.pygame.close()


    def clone(self):
        """Creates a new instance of the enviroment, which is a copy of the current state.
        Returns:
            clone: clone of the current environment.
        """
        # make a fresh env with the same players and no render
        clone = BriscolaEnv([p.name for p in self.players], render_mode=None)

        clone.n_players = self.n_players
        clone.done = self.done
        for i in range(len(clone.players)):
            real_player = self.players[i]
            clone.players[i].hand = [card for card in real_player.hand]
            for card in clone.players[i].hand:
                if hasattr(card, "image"):
                    del card.image
        clone.played_cards_history = list(self.played_cards_history)
        for card in clone.played_cards_history:
            if hasattr(card, "image"):
                del card.image
        clone.turn_history = list(self.turn_history)
        clone.current_player_id = self.current_player_id
        clone.turn_order = list(self.turn_order)
        clone.turn_counter = self.turn_counter
        clone.cards_on_table = list(self.cards_on_table)
        for card in clone.cards_on_table:
            if hasattr(card, "image"):
                del card.image

        # clone dealer
        clone.dealer = self.dealer.clone()

        # clone players (but not their agents/rendering)
        clone.players = []
        for orig in self.players:
            new_p = BriscolaPlayer(orig.name)
            new_p.hand = list(orig.hand)
            new_p.points = orig.points
            new_p.taken_cards = list(orig.taken_cards)
            clone.players.append(new_p)

        # briscola info
        clone.briscola_spy = self.briscola_spy
        clone.briscola_suit_id = self.briscola_suit_id

        return clone

    
    def clone_from_observation(self):
        """Creates a copy of the environment from the side of one player, so randomly shuffling remaining unknown cards.
        Returns:
            obs_clone: random clone of the environment from point of view of one player.
        """
        obs = self.get_observation()
        clone = self.clone()

        # mix deck and other player hand
        if self.dealer.cards_left > 0: # if deck is empty, don't shuffle because other player hand is known
            hand_sizes = []
            briscola_spy = clone.dealer.deck.pop(0)
            for player_id in range(len(clone.players)):
                hand_sizes.append(len(clone.players[player_id].hand))
                if player_id == clone.current_player_id:
                    continue
                for _ in range(len(clone.players[player_id].hand)):
                    card = clone.players[player_id].hand.pop()
                    clone.dealer.deck.append(card)
            clone.dealer.shuffle()
            clone.dealer.deck.insert(0, briscola_spy)
            for player_id in range(len(clone.players)):
                if player_id == clone.current_player_id:
                    continue
                for _ in range(hand_sizes[player_id]):
                    clone.players[player_id].hand.append(clone.dealer.deal())
            
        clone.render_mode = None
        return clone

File: cardroom/briscola/src/env/env.rs
use super::cards::{BriscolaCard, BriscolaSuit};
use super::dealer::BriscolaDealer;
use super::player::BriscolaPlayer;
use super::player::BriscolaAction;
use super::utils::env_from_obs;

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum BriscolaResult {
    P0Win,
    Draw,
    P1Win,
    InProgress,
}

impl BriscolaResult {
    pub fn winner_idx(&self) -> Option<usize> {
        match self {
            BriscolaResult::P0Win => Some(0),
            BriscolaResult::P1Win => Some(1),
            _ => None,
        }
    }
}

#[derive(Clone)]
pub struct BriscolaState {
    // describes the state of the game with perfect information
    pub n_players: usize,
    pub players: Vec<BriscolaPlayer>,
    pub done: bool,
    pub played_cards_history: Vec<BriscolaCard>,
    pub turn_history: Vec<usize>,
    pub dealer: BriscolaDealer,
    pub cards_on_table: Vec<BriscolaCard>,
    pub current_player_id: usize,
    pub turn_order: Vec<usize>,
    pub turn_counter: u8,
    pub result: BriscolaResult,
}

#[derive(Clone)]
pub struct BriscolaObs {
    // describes the state of the from the point of view of a player, imperfect information
    pub n_players: usize,
    pub player: BriscolaPlayer,
    pub points: Vec<i32>,
    pub done: bool,
    pub played_cards_history: Vec<BriscolaCard>,
    pub turn_history: Vec<usize>,
    pub briscola: BriscolaSuit,
    pub briscola_spy: Option<BriscolaCard>,
    pub cards_in_deck: usize,
    pub cards_on_table: Vec<BriscolaCard>,
    pub current_player_id: usize,
    pub turn_order: Vec<usize>,
    pub turn_counter: u8,
    pub result: BriscolaResult,
}
#[derive(Clone)]
pub struct BriscolaEnv {
    pub n_players: usize,
    pub players: Vec<BriscolaPlayer>,
    pub done: bool,
    pub played_cards_history: Vec<BriscolaCard>,
    pub turn_history: Vec<usize>,
    pub dealer: BriscolaDealer,
    pub cards_on_table: Vec<BriscolaCard>,
    pub current_player_id: usize,
    pub turn_order: Vec<usize>,
    pub turn_counter: u8,
    pub result: BriscolaResult,
}

impl BriscolaEnv {
    pub fn new() -> Self {
        let n_players = 2;
        let names: Vec<String> = (0..n_players).map(|i| format!("Player_{}", i)).collect();
        Self::new_with_names(names)
    }

    pub fn new_with_names(names: Vec<String>) -> Self {
        let n_players = names.len();
        let done = false;
        let mut players = vec![];
        let dealer = BriscolaDealer::new();
        for name in names {
            players.push(BriscolaPlayer::new(name));
        }
        let cards_on_table = vec![];
        let played_cards_history = vec![];
        let turn_history = vec![];
        let current_player_id = 0;
        let turn_order: Vec<usize> = (0..n_players).collect();
        let turn_counter = 1;
        let result = BriscolaResult::InProgress;

        let mut env = Self{n_players, players, done, played_cards_history, turn_history, dealer, cards_on_table, current_player_id, turn_order, turn_counter, result};
        env.reset();
        env
    }

    pub fn reset_with_deck(&mut self, deck: Vec<BriscolaCard>) {
        self.done = false;
        self.played_cards_history = vec![];
        self.turn_history = vec![];
        self.dealer = BriscolaDealer::from_deck(deck);
        self.result = BriscolaResult::InProgress;
        for player in &mut self.players {
            player.reset();
        }
        self.current_player_id = 0;
        self.set_turn_order(0);
        for &id in &self.turn_order {
            let player = &mut self.players[id];
            for _ in 0..3 {
                player.hand.push(self.dealer.deal());
            }
        }
        self.turn_counter = 1;
    }

    pub fn reset(&mut self) {
        self.done = false;
        self.played_cards_history = vec![];
        self.turn_history = vec![];
        self.dealer = BriscolaDealer::new();
        self.result = BriscolaResult::InProgress;

        for player in &mut self.players {
            player.reset();
        }
        // draw first player
        self.current_player_id = rand::random_range(0..self.n_players);
        self.set_turn_order(self.current_player_id);
        // deal first 3 cards
        for &id in &self.turn_order {
            let player = &mut self.players[id];
            for _ in 0..3 {
                player.hand.push(self.dealer.deal());
            }
        }

        self.turn_counter = 1;
    }

    pub fn set_turn_order(&mut self, first_id: usize) {
        self.turn_order.clear();
        for i in 0..self.n_players {
            self.turn_order.push((first_id + i) % self.n_players);
        }
    }

    pub fn pass_turn(&mut self) {
        self.current_player_id = (self.current_player_id + 1) % self.n_players
    }

    pub fn resolve_turn(&mut self) {
        let first_suit = self.cards_on_table[0].suit;
        let mut played_points = 0;
        let mut winning_idx = 0;
        let mut winning_rank = i32::MIN;

        for (i, card) in self.cards_on_table.iter().enumerate() {
            let effective_rank = card.rank + if card.suit == first_suit { 50 } else { 0 };
            played_points += card.points;
            if effective_rank > winning_rank {
                winning_rank = effective_rank;
                winning_idx = i;
            }
        }

        let winning_player_id = self.turn_order[winning_idx];
        self.players[winning_player_id].points += played_points;
        self.players[winning_player_id].taken_cards.extend(self.cards_on_table.drain(..));
        self.set_turn_order(winning_player_id);
        self.current_player_id = winning_player_id;
    }

    pub fn step(&mut self, action: BriscolaAction) {
        // executes action of current player and updates the game
        let is_last = self.current_player_id == *self.turn_order.last().unwrap();
        
        let card = self.players[self.current_player_id].play_card(action);
        self.played_cards_history.push(card);
        self.cards_on_table.push(card);
        self.turn_history.push(self.current_player_id);

        if !is_last {self.pass_turn();}
        else {
            self.resolve_turn();
            self.turn_counter += 1;

            // check if game is over, else deal new cards
            if self.turn_counter > 20 {
                let (result, _points) = self.check_result();
                self.result = result;
                self.done = true;
            }
            else if self.dealer.deck.len() > 0 {
                for id in &self.turn_order {
                    self.players[*id].hand.push(self.dealer.deal());
                }
            }
        }
    }

    pub fn get_state(&self) -> BriscolaState {
        BriscolaState {
            n_players: self.n_players,
            players: self.players.clone(),
            done: self.done,
            played_cards_history: self.played_cards_history.clone(),
            turn_history: self.turn_history.clone(),
            dealer: self.dealer.clone(),
            cards_on_table: self.cards_on_table.clone(),
            current_player_id: self.current_player_id,
            turn_order: self.turn_order.clone(),
            turn_counter: self.turn_counter,
            result: self.result,
        }
    }

    pub fn get_obs(&self) -> BriscolaObs {
        BriscolaObs {
            n_players: self.n_players,
            player: self.players[self.current_player_id].clone(),
            points: vec![self.players[0].points, self.players[1].points],
            done: self.done,
            played_cards_history: self.played_cards_history.clone(),
            turn_history: self.turn_history.clone(),
            briscola: self.dealer.briscola,
            briscola_spy: self.dealer.deck.first().cloned(),
            cards_in_deck: self.dealer.deck.len(),
            cards_on_table: self.cards_on_table.clone(),
            current_player_id: self.current_player_id,
            turn_order: self.turn_order.clone(),
            turn_counter: self.turn_counter,
            result: self.result,
        }
    }

    pub fn check_result(&mut self) -> (BriscolaResult, Vec<i32>) {
        // checks result of the game
        let mut points = vec![];
        for player in &self.players {
            points.push(player.points);
        }
        let result = 
            if points[0] == points[1] {BriscolaResult::Draw}
            else if points[0] > points[1] {BriscolaResult::P0Win}
            else {BriscolaResult::P1Win};

        (result, points)
    }

    pub fn get_legal_actions(&self) -> Vec<BriscolaAction> {
        let player = &self.players[self.current_player_id];
        player.get_legal_actions()
    }

    pub fn render(&self) {
        // text rendering of the game
        // pre-game text
        if self.turn_counter == 1 && self.cards_on_table.len() == 0 {
            println!("==================");
            println!("Starting new game");
            println!("==================");
            println!("Players: {}, {}", self.players[0].name, self.players[1].name);
            println!("Starting player: {}", self.players[self.current_player_id].name);
            println!("Briscola suit: {} - spy: {}", self.dealer.briscola.name(), self.dealer.deck[0].name());
            return;
        }
        // pre-round text (render happens after step)
        if self.cards_on_table.len() == 1 {
            println!("** Turn {} - {} first of hand **", self.turn_counter, self.players[*self.turn_history.last().unwrap()].name);
            println!("{} plays {}", self.players[*self.turn_history.last().expect("No turns played.")].name, self.played_cards_history.last().unwrap().name());
        }
        //post-round text
        else {
            println!("{} plays {}", self.players[*self.turn_history.last().expect("No turns played.")].name, self.played_cards_history.last().unwrap().name());
        }

        //post-game text
        if self.done {
            println!("** Game finished **");
            for player in &self.players {
                println!("{}: {} points", player.name, player.points);
            }
            if self.result == BriscolaResult::Draw {
                println!("*** Draw ***");
            } else {
                println!("*** {} wins ***", self.players[self.result.winner_idx().unwrap()].name);
            }
        }

    }

    pub fn random_clone(&self) -> Self{
        //returns a clone fo the env generated with imperfect information from the perspective of the current player.
        // opponent hand and deck are randomized
        env_from_obs(&self.get_obs())
    }
}


#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    fn play_game() -> BriscolaEnv {
        // plays simulated game where both players always play card 0
        let mut env = BriscolaEnv::new();

        while !env.done {
            env.step(BriscolaAction::Card0);
        }
        env
    }

    #[test]
    fn test_game() {
        for _i in 0..100 {
            let env = play_game();

            assert!(env.done == true);
            assert!(env.players.iter().map(|p| p.points).sum::<i32>() == 120);
            assert!(env.turn_history.len() == 40);
            assert!(env.played_cards_history.len() == 40);
            assert!(vec![BriscolaResult::P0Win, BriscolaResult::Draw, BriscolaResult::P1Win].contains(&env.result));
        }
    }

    #[test]
    fn test_reset() {
        let mut env = BriscolaEnv::new();
        assert!(env.done == false);
        assert!(env.players.iter().map(|p| p.points).sum::<i32>() == 0);
        assert!(env.turn_counter == 1);
        assert!(env.turn_history.len() == 0);
        assert!(env.played_cards_history.len() == 0);
    
        env = play_game();
        env.reset();
        assert!(env.done == false);
        assert!(env.players.iter().map(|p| p.points).sum::<i32>() == 0);
        assert!(env.turn_counter == 1);
        assert!(env.turn_history.len() == 0);
        assert!(env.played_cards_history.len() == 0);
    }

    #[rstest]
    #[case(0, 1)]
    #[case(1, 0)]
    fn test_set_turn_order(#[case] current_player_id: usize, #[case] expected: usize) {
        let mut env = BriscolaEnv::new();
        env.current_player_id = current_player_id;
        env.pass_turn();

        assert_eq!(env.current_player_id, expected);
    }

    fn play_steps(env: &mut BriscolaEnv, n: usize) {
        for _ in 0..n {
            if !env.done { env.step(BriscolaAction::Card0); }
        }
    }

    #[rstest]
    #[case(0)]
    #[case(5)]
    #[case(10)]
    #[case(15)]
    fn test_clone(#[case] steps: usize) {
        let mut env = BriscolaEnv::new();
        play_steps(&mut env, steps);

        let mut clone = env.clone();

        while !env.done {
            env.step(BriscolaAction::Card0);
            clone.step(BriscolaAction::Card0);
            assert_eq!(env.turn_counter, clone.turn_counter);
            assert_eq!(env.current_player_id, clone.current_player_id);
            assert_eq!(env.cards_on_table, clone.cards_on_table);
            assert_eq!(env.players[0].points, clone.players[0].points);
            assert_eq!(env.players[1].points, clone.players[1].points);
        }
        assert_eq!(env.result, clone.result);
    }

    #[rstest]
    #[case(0)]
    #[case(5)]
    #[case(10)]
    #[case(15)]
    #[case(35)]
    #[case(37)]
    #[case(39)]
    fn test_random_clone_observable_state(#[case] steps: usize) {
        let mut env = BriscolaEnv::new();
        play_steps(&mut env, steps);

        let obs = env.get_obs();
        let clone_obs = env.random_clone().get_obs();

        assert_eq!(obs.player.hand, clone_obs.player.hand);
        assert_eq!(obs.cards_on_table, clone_obs.cards_on_table);
        assert_eq!(obs.played_cards_history, clone_obs.played_cards_history);
        assert_eq!(obs.points, clone_obs.points);
        assert_eq!(obs.briscola, clone_obs.briscola);
        assert_eq!(obs.briscola_spy, clone_obs.briscola_spy);
        assert_eq!(obs.cards_in_deck, clone_obs.cards_in_deck);
        assert_eq!(obs.current_player_id, clone_obs.current_player_id);
        assert_eq!(obs.turn_order, clone_obs.turn_order);
        assert_eq!(obs.turn_counter, clone_obs.turn_counter);
    }

    #[rstest]
    #[case(0)]
    #[case(5)]
    #[case(10)]
    #[case(15)]
    #[case(35)]
    #[case(37)]
    #[case(39)]
    fn test_random_clone_card_uniqueness(#[case] steps: usize) {
        use std::collections::HashSet;
        use crate::env::cards::{BriscolaFace, BriscolaSuit};

        let mut env = BriscolaEnv::new();
        play_steps(&mut env, steps);

        let clone = env.random_clone();

        let total = clone.played_cards_history.len()
            + clone.dealer.deck.len()
            + clone.players.iter().map(|p| p.hand.len()).sum::<usize>();
        assert_eq!(total, 40);

        let mut seen: HashSet<(BriscolaFace, BriscolaSuit)> = HashSet::new();
        for card in &clone.played_cards_history {
            assert!(seen.insert((card.face, card.suit)), "duplicate in history");
        }
        for card in &clone.dealer.deck {
            assert!(seen.insert((card.face, card.suit)), "duplicate in deck");
        }
        for player in &clone.players {
            for card in &player.hand {
                assert!(seen.insert((card.face, card.suit)), "duplicate in hand");
            }
        }

        if let Some(spy) = env.get_obs().briscola_spy {
            assert_eq!(clone.dealer.deck[0], spy);
        } else {
            assert!(clone.dealer.deck.is_empty());
        }
    }

    /// When the deck is empty the random clone's opponent hand is fully determined
    /// and must match the exact clone's opponent hand (same card set).
    #[rstest]
    #[case(35)]
    #[case(37)]
    #[case(39)]
    fn test_random_clone_empty_deck_same_as_clone(#[case] steps: usize) {
        use crate::env::cards::{BriscolaFace, BriscolaSuit};

        let mut env = BriscolaEnv::new();
        play_steps(&mut env, steps);

        let opponent_id = 1 - env.current_player_id;
        let exact = env.clone();
        let random = env.random_clone();

        let mut exact_hand: Vec<(BriscolaFace, BriscolaSuit)> = exact.players[opponent_id].hand
            .iter().map(|c| (c.face, c.suit)).collect();
        let mut random_hand: Vec<(BriscolaFace, BriscolaSuit)> = random.players[opponent_id].hand
            .iter().map(|c| (c.face, c.suit)).collect();
        exact_hand.sort_by_key(|&(f, s)| (f as u8, s as u8));
        random_hand.sort_by_key(|&(f, s)| (f as u8, s as u8));
        assert_eq!(exact_hand, random_hand);
    }

    #[test]
    fn test_random_clone_game_completes() {
        for _ in 0..100 {
            let env = BriscolaEnv::new();
            let mut clone = env.random_clone();

            while !clone.done { clone.step(BriscolaAction::Card0); }

            assert_eq!(clone.players.iter().map(|p| p.points).sum::<i32>(), 120);
            assert!(vec![BriscolaResult::P0Win, BriscolaResult::Draw, BriscolaResult::P1Win]
                .contains(&clone.result));
        }
    }
}